doc/dev/radosgw/virtual_hosting.md

```diff
@@ -0,0 +1,120 @@
+
+## motivation
+
+radosgw currently serves all rest apis from the same namespace. some are hosted at the root, with others under paths like `admin`, `auth`, `swift` etc. most apis are enabled by default, so parts of the root s3 namespace are unavailable because those paths get routed to other apis
+
+s3 shares the root path with several other apis (iam, s3control, sts, sns) and relies on tricks to disambiguate the requests and route them correctly. these tricks are based on assumptions that may not be reliable, especially as the apis evolve
+
+to avoid these conflicts in path-based routing, we can first route requests by hostname
+
+this _could_ be accomplished by deploying separate radosgw processes for each api and configuring dns accordingly. however,
+* radosgw processes are heavy-weight, having several background threads and a main pool of 512 threads
+* the overall load would not be well-balanced over radosgws, with some apis expecting very little traffic
+* the metadata cache broadcasts invalidation messages, so will limit scaling to many radosgws
+
+radosgw can continue serving all of these apis from the same frontend socket by instead routing the requests according to their http `Host` header. this [virtual hosting](https://en.wikipedia.org/wiki/Virtual_hosting) technique is common in web servers
+
+## existing request routing
+
+config options:
+
+```
+rgw_enable_apis = s3, s3control, s3website, swift, swift_auth, admin, sts, iam, notifications
+rgw_dns_name =
+rgw_dns_s3website_name =
+```
+
+request flow:
+
+1. `process_request()` calls `get_manager()` on the root `RGWRESTMgr` to route the given request path to its final `RGWRESTMgr`
+2. `RGWRESTMgr::get_handler()` creates the request `RGWHandler` for that api
+3. `RGWHandler::get_op()` creates the `RGWOp` based on request method
+
+during startup, `rgw::AppMain::cond_init_apis()` builds a tree of `RGWRESTMgr`s to match nested request paths
+
+`RGWRESTMgr::register_resource(string resource, RGWRESTMgr* mgr)` is used to register child managers at relative paths for static routing
+
+each api listed in `rgw_enable_apis` registers their own `RGWRESTMgr` at the configured path. ex `RGWRESTMgr_SWIFT` at `rgw_swift_url_prefix`, `RGWRESTMgr_Admin` at `rgw_admin_entry`
+
+s3 paths can have arbitrary bucket names, so require dynamic routing. `RGWRESTMgr_S3` is registered as the default manager, selected when no registered static paths match. `RGWRESTMgr_S3::get_handler()` interpets the path itself when producing the request handler
+
+other apis hosted at the root path (iam, s3control, s3website, sns, sts) must be routed internally by `RGWRESTMgr_S3`
+
+the `Host` header is matched against `rgw_dns_name`/`rgw_dns_s3website_name` to:
+* disambiguate between s3 and s3website requests, and
+* transform s3 virtual-host style requests into path-style requests
+
+## virtual host request routing
+
+new config options:
+
+```
+rgw_dns_admin_name =
```

Comment:
We might also need to consider decoupling endpoints for admin and S3 in multisite (period config), as if one utilizes this one, multisite requests might fail due to the missing admin/S3 handler by that hostname.

Reply:
true, the multisite config stores a list of zone/zonegroup "endpoints" which we assume will work for both s3 and /admin requests. probably need to document that multisite configurations must leave rgw_dns_admin_name empty